### PR TITLE
feat(smart-contracts): add `onKeyExtend` hook

### LINKED ISSUE #9589: onExtendKey hook
Let's add a hook when a key is renewed or extended!

### PR DESCRIPTION
**Pull request by clemsos**
# Description

Adds support for a `onKeyExtend` hook

# Issues

Fixes #9589
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/interfaces/hooks/ILockKeyExtendHook.sol [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  /**
+   * @notice This hook every time a key is extended.
+   * @param tokenId tje id of the key
+   * @param from the msg.sender making the purchase
+   * @param newTimestamp the account which referred this key sale
+   */
+  function onKeyExtend(
+    uint tokenId,
+    address from,
+    uint newTimestamp
```
Comment (julien51) [inline smart-contracts/contracts/interfaces/hooks/ILockKeyExtendHook.sol:20]: Do we want to also show the old timestamp?

File: smart-contracts/contracts/mixins/MixinKeys.sol [hunk 2]
```diff
@@ -271,6 +271,15 @@ contract MixinKeys is
     _keys[_tokenId].expirationTimestamp = newTimestamp;
 
     emit KeyExtended(_tokenId, newTimestamp);
+
+    // call the hook
+    if(address(onKeyExtendHook) != address(0)) {
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinKeys.sol:276]: I think this will be called for `extend`, `grantKeyExtension` and `renew`, correct?

**Review by julien51 — APPROVED** [review]
Can we add tests to make sure it is called in all the functions that extend the key duration?

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (clemsos) [reply smart-contracts/contracts/mixins/MixinKeys.sol:276]: yes, the three calls it to extend the time o the key

File: smart-contracts/contracts/interfaces/hooks/ILockKeyExtendHook.sol (hunk 1 above)
Comment (clemsos) [reply smart-contracts/contracts/interfaces/hooks/ILockKeyExtendHook.sol]: ok, added

**Comment by clemsos** [comment]
> Can we add tests to make sure it is called in all the functions that extend the key duration?

yep added

**Pull request merged**